src/mlpack/bindings/R/print_input_processing.hpp

```diff
@@ -0,0 +1,209 @@
+    const typename boost::disable_if<arma::is_arma_type<T>>::type* = 0,
+    const typename boost::disable_if<data::HasSerialize<T>>::type* = 0,
+    const typename boost::disable_if<std::is_same<T,
+        std::tuple<data::DatasetInfo, arma::mat>>>::type* = 0)
+{
+  if (!d.required)
+  {
+    /**
+     * This gives us code like:
+     *
+     *  if (!identical(\<param_name\>, NA)) {
+     *     CLI_SetParam<type>("\<param_name\>", \<param_name\>)
+     *  }
+     */
+    MLPACK_COUT_STREAM << "  if (!identical(" << d.name ;
+    if (d.cppType == "bool")
+    {
+      MLPACK_COUT_STREAM << ", FALSE)) {" << std::endl;
+    }
+    else
+    {
+      MLPACK_COUT_STREAM << ", NA)) {" << std::endl;
+    }
+    MLPACK_COUT_STREAM << "    CLI_SetParam" << GetType<T>(d) << "(\""
+        << d.name << "\", " << d.name << ")" << std::endl;
+    MLPACK_COUT_STREAM << "  }" << std::endl; // Closing brace.
+  }
+  else
+  {
+    /**
+     * This gives us code like:
+     *
+     *     CLI_SetParam<type>("\<param_name\>", \<param_name\>)
+     */
+    MLPACK_COUT_STREAM << "  CLI_SetParam" << GetType<T>(d) << "(\""
+              << d.name << "\", " << d.name << ")" << std::endl;
+  }
+  MLPACK_COUT_STREAM << std::endl; // Extra line is to clear up the code a bit.
+}
+
+/**
+ * Print input processing for a matrix type.
+ */
+template<typename T>
+void PrintInputProcessing(
+    const util::ParamData& d,
+    const typename boost::enable_if<arma::is_arma_type<T>>::type* = 0)
+{
+  if (!d.required)
+  {
+    /**
+     * This gives us code like:
+     *
+     *  if (!identical(\<param_name\>, NA)) {
+     *     CLI_SetParam<type>("\<param_name\>", to_matrix(\<param_name\>))
+     *  }
+     */
+    MLPACK_COUT_STREAM << "  if (!identical(" << d.name << ", NA)) {"
+        << std::endl;
+    MLPACK_COUT_STREAM << "    CLI_SetParam" << GetType<T>(d) << "(\""
+        << d.name << "\", to_matrix(" << d.name << "))" << std::endl;
+    MLPACK_COUT_STREAM << "  }" << std::endl; // Closing brace.
+  }
+  else
+  {    /**
+     * This gives us code like:
+     *
+     *     CLI_SetParam<type>("\<param_name\>", to_matrix(\<param_name\>))
+     */
+    MLPACK_COUT_STREAM << "  CLI_SetParam" << GetType<T>(d) << "(\""
+              << d.name << "\", to_matrix(" << d.name << "))" << std::endl;
+  }
+  MLPACK_COUT_STREAM << std::endl; // Extra line is to clear up the code a bit.
+}
+
+/**
+ * Print input processing for a matrix with info type.
+ */
+template<typename T>
+void PrintInputProcessing(
+    const util::ParamData& d,
+    const typename boost::enable_if<std::is_same<T,
+        std::tuple<data::DatasetInfo, arma::mat>>>::type* = 0)
+{
+  if (!d.required)
+  {
+    /**
+     * This gives us code like:
+     *
+     *  if (!identical(\<param_name\>, NA)) {
+     *     \<param_name\> = to_matrix_with_info(\<param_name\>)
+     *     CLI_SetParam\<type\>("\<param_name\>", \<param_name\>$info,
+     *                         \<param_name\>$data)
```

Comment:
This maybe is a question for @coatless and @eddelbuettel too.  I know that R does not specify types for function arguments (or, well, I think!), but we should probably have a discussion on what's expected to be passed, especially for the type the code here uses: the "categorical matrix".

In mlpack we support categorical features basically by pairing a matrix with a vector of `bool`s that indicate which dimensions are categorical and which are numeric.  We have some nice structures that make that a little bit easier to work with, but when we interface with other languages, what we have to pass between languages amounts to a `bool` array and a matrix.

That part itself isn't hard (the code here already does it), but the more difficult part that involves a bit of "taste" is what the user should pass in to the R function.  Should we expect them to pass in what this code expects?  (I think that is a data frame with a column `$info` and a column (columns?) `$data`.)  Does that feel "R native"?  Or can we do better?  I really have no idea, I'm just posing the question. :)

Reply:
R is dynamically typed so ... kinda hard to impose a type.

R has a specific type for categorical variables: `factor`.  Essentially an `int` vector plus a string hash map.  So `{blue, green, blue, red}` could be decoded as {`1,2,1,3}` with the associate vector {"blue", "green", "red"} for "labels" for (1:3).  Sadly that `factor` type does not travel too well to/from C++.

Reply:
I see---so, does it make sense to try and take a `factor` as inputs to our bindings, then do a little bit of mapping?  All we really need to pass to the mlpack side is that `int` vector; we can leave the string hash map on the R side.  I am not sure of the exact representation, but if we can wrangle whatever it is we get as input into an Armadillo matrix plus a boolean vector (which would take value `false` for any "normal" numeric dimensions and `true` for any `factor` dimensions), then I think that is all we'd need to do.